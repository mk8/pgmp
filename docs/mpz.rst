`!mpz` data type
================

The `!mpz` data type can store integer numbers of arbitrary size, up to the
maximum memory allowed by PostgreSQL data types (about 1GB).

Every PostgreSQL integer type (`!int16`, `!int32`, `!int64`) can be converted
to `!mpz`. Not integer types (`!float4`, `!float8`, `!numeric`) are truncated.
Cast from integer types are automatic, whereas non integer require explicit
cast (but are implicitly converted in assignment).

.. code-block:: sql

    SELECT mpz(10000);      -- Cast using mpz as a function
    SELECT 10000::mpz;      -- PostgreSQL-style cast

.. todo:: Not integer input

`!mpz` values can be compared using the regular PostgreSQL comparison
operators. Indexes on `!mpz` columns can be created using the *btree* method.


`!mpz` textual input/output
---------------------------

.. function:: mpz(text)
.. function:: mpz(text, base)

    Convert a textual representation into an `!mpz` number. The form
    :samp:`{text}::mpz` is equivalent to :samp:`mpz({text})`.

    White space is allowed in the string, and is simply ignored.

    The *base* may vary from 2 to 62, or if *base* is 0 or omitted, then the
    leading characters are used: ``0x`` and ``0X`` for hexadecimal, ``0b`` and
    ``0B`` for binary, ``0`` for octal, or decimal otherwise.

    For bases up to 36, case is ignored; upper-case and lower-case letters
    have the same value.  For bases 37 to 62, upper-case letter represent the
    usual 10..35 while lower-case letter represent 36..61.

    .. code-block:: sql

        =# SELECT '0x10'::mpz AS hex, '10'::mpz AS dec , '010'::mpz AS oct, '0b10'::mpz AS bin;
         hex | dec | oct | bin
        -----+-----+-----+-----
         16  | 10  | 8   | 2

.. todo:: I should write a psql pygmets lexer

.. todo:: Handle base 0 in mpz(str, int)

.. function:: text(z)

.. function:: text(z, base)

    Convert the `!mpz` *z* into a string. The form :samp:`{z}::text` is
    equivalent to :samp:`text({z})`.

    *base* may vary from 2 to 62 or from −2 to −36.  For base in the range
    2..36, digits and lower-case letters are used; for −2..−36, digits and
    upper-case letters are used; for 37..62, digits, upper-case letters, and
    lower-case letters (in that significance order) are used. If *base* is not
    specified, 10 is assumed.

.. todo::
    Not sure if handling 0 correctly here too, or if it should handled at all

.. todo:: write a wrap() function


Arithmetic Operators and Functions
----------------------------------

These operators can either work on `!mpz` arguments or take an integer
argument that will be implicitly converted.  Operators taking a :math:`2^n`
argument always use an integer as right argument.

A few operators may use optimized algorithms when one of the arguments is an
integer.

.. note::
    GMP defines many structures in terms of `!long` or `!unsigned long`, whose
    definitions may vary across platforms. PostgreSQL instead offers data
    types with a defined number of bytes (e.g. `!int4`, `!int8`). For this
    reason, functions taking an integer as argument are defined as `!int8`,
    but they may actually fail if the server is 32 bit and the argument don't
    fit into an `!int4`.

.. table:: Arithmetic operators

    =========== =============================== =================== ===========
    Operator    Description                     Example             Return
    =========== =============================== =================== ===========
    `!-`        Unary minus                     `!- 5::mpz`         -5
    `!+`        Unary plus                      `!+ 5::mpz`         5
    `!+`        Addition                        `!2::mpz + 3::mpz`  5
    `!-`        Subtraction                     `!2::mpz - 3::mpz`  -1
    `!*`        Multiplication                  `!7::mpz * 3::mpz`  21
    `!<<`       Multiplication by :math:`2^n`   `!3::mpz << 2`      12
    `!^`        Power (1)                       `!3::mpz ^ 2`       9
    =========== =============================== =================== ===========

Notes:

(1)
    See also the `exponentiation functions`_.


.. function:: abs(z)

    Return the absolute value of *z*.


Division Operators and Functions
--------------------------------

.. table:: Division operators

    =========== =============================== ==================== =======
    Operator    Description                     Example              Return
    =========== =============================== ==================== =======
    `!/`        Division quotient               `!7::mpz / 3::mpz`   2

                Rounding towards zero           `!-7::mpz / 3::mpz`  -2

    `!%`        Division reminder               `!7::mpz % 3::mpz`   1

                Rounding towards zero           `!-7::mpz % 3::mpz`  -1

    `+/`        Division quotient               `!7::mpz +/ 3::mpz`  3

                Rounding towards +infinity      `!-7::mpz +/ 3::mpz` -2

    `+%`        Division reminder               `!7::mpz +% 3::mpz`  -2

                Rounding towards +infinity      `!-7::mpz +% 3::mpz` -1

    `!-/`       Division quotient               `!7::mpz -/ 3::mpz`  2

                Rounding towards -infinity      `!-7::mpz -/ 3::mpz` -3

    `!-%`       Division reminder               `!7::mpz -% 3::mpz`  1

                Rounding towards -infinity      `!-7::mpz -% 3::mpz` 2

    `/!`        Exact division (1)              `!21::mpz /! 7::mpz` 3
    =========== =============================== ==================== =======

Notes:

(1)
    The exact division operator (`!/!`) produce correct results only when it
    is known in advance that :math:`d` divides :math:`n`.  The operator is
    much faster than the other division operators, and is the best choice when
    exact division is known to occur, for example reducing a rational to
    lowest terms.


.. table:: Division operators for powers of 2

    ======== ==================================== =================== =======
    Operator Description                          Example             Return
    ======== ==================================== =================== =======
    `!>>`    Quotient of division by :math:`2^n`  `!1027::mpz >> 3`   128

             Rounding towards zero                `!-1027::mpz >> 3`  -128

    `!%>`    Remainder of division by :math:`2^n` `!1027::mpz %> 3`   3

             Rounding towards zero                `!-1027::mpz %> 3`  -3

    `!+>>`   Quotient of division by :math:`2^n`  `!1027::mpz +>> 3`  129

             Rounding towards +infinity           `!-1027::mpz +>> 3` -128

    `!+%>`   Remainder of division by :math:`2^n` `!1027::mpz +%> 3`  -5

             Rounding towards +infinity           `!-1027::mpz +%> 3` -3

    `!->>`   Quotient of division by :math:`2^n`  `!1027::mpz ->> 3`  128

             Rounding towards -infinity           `!-1027::mpz ->> 3` -129

    `!-%>`   Remainder of division by :math:`2^n` `!1027::mpz -%> 3`  3

             Rounding towards -infinity           `!-1027::mpz -%> 3` 5
    ======== ==================================== =================== =======

For all the division-related operators :math:`n \oslash d`, :math:`q` and
:math:`r` will satisfy :math:`n = q \cdot d + r`, and :math:`r` will satisfy
:math:`0 \le |r| \lt |d|`.

.. note::
    Only the truncating division and reminder (`!/` and `!%`) have the correct
    precedence respect to addition, subtraction and multiplication.
    See `the PostgreSQL precedence table`__ for further details.

    .. __: http://www.postgresql.org/docs/9.0/static/sql-syntax-lexical.html#SQL-PRECEDENCE-TABLE


.. function:: divisible(n, d)

.. function:: divisible_2exp(n, b)

    Return `!true` if *n* is exactly divisible by *d*, or in the case of
    `!divisible_2exp()` by :math:`2^b`.

    :math:`n` is divisible by :math:`d` if there exists an integer :math:`q`
    satisfying :math:`n = q \cdot d`.  Unlike the other division operators,
    *d*\=0 is accepted and following the rule it can be seen that only 0
    is considered divisible by 0.


.. function:: congruent(n, c, d)

.. function:: congruent_2exp(n, c, b)

    Return `!true` if *n* is congruent to *c* modulo *d*, or in the case of
    `!congruent_2exp()` modulo :math:`2^b`.

    :math:`n` is congruent to :math:`c \mod d` if there exists an integer
    :math:`q` satisfying :math:`n = c + q \cdot d`. Unlike the other division
    operators, *d*\=0 is accepted and following the rule it can be seen that n
    and c are considered congruent mod 0 only when exactly equal.

    .. todo:: functions `divisible()` and `congruent()`


.. todo:: integer fast path

.. todo::
    fast path on int64 for 64 bit backends? Maybe introduce a long data type?


Exponentiation Functions
------------------------

.. function:: pow(base, exp)

    Return *base* raised to *exp*.

    *exp* is defined as `!int8` but must fit into a `!long` as defined on the
    server.

    The function is an alias for the `!^` operator.


.. function:: powm(base, exp, mod)

    Return (*base* raised to *exp*) modulo *mod*.

    Negative *exp* is supported if an inverse *base^-1* mod *mod* exists (see
    `invert()` function). If an inverse doesn't exist then a divide by zero is
    raised.


.. function:: powm_sec(base, exp, mod)

    Return (*base* raised to *exp*) modulo *mod*.

    It is required that *exp* > 0 and that *mod* is odd.

    This function is designed to take the same time and have the same cache
    access patterns for any two same-size arguments, assuming that function
    arguments are placed at the same position and that the machine state is
    identical upon function entry. This function is intended for cryptographic
    purposes, where resilience to side-channel attacks is desired.

.. todo:: `powm()`, `powm_sec()`.


Root Extraction Functions
-------------------------

.. function:: root(op, n)

    Return the truncated integer part of the *n*\th root of *op*.

    *n* is defined as `!int8` but must fit into a `!long` as defined on the
    server.

.. function:: rootrem(op, n)

    Return a tuple of 2 elements with the truncated integer part of the *n*\th
    root of *op* and the remainder (*i.e.* *op* - *root* ^ *n*).

    .. code-block:: sql

        =# select * from rootrem(28, 3);
         root | rem
        ------+-----
         3    | 1

.. function:: sqrt(op)

    Return the truncated integer part of the square root of *op*.

.. function:: sqrtrem(op)

    Return a tuple of 2 elements with the truncated integer part of the square
    root of *op* and the remainder (*i.e.* *op* - *root* \* *root*).

    .. code-block:: sql

        =# select * from sqrtrem(83);
         root | rem
        ------+-----
         9    | 2

.. function:: perfect_power(op)

    Return ``true`` if *op* is a perfect power, *i.e.*, if there exist
    integers :math:`a` and :math:`b`, with :math:`b>1`, such that *op* equals
    :math:`a^b`.

    Under this definition both 0 and 1 are considered to be perfect powers.
    Negative values of op are accepted, but of course can only be odd perfect
    powers.

.. function:: perfect_square(op)

    Return ``true`` if *op* is a perfect square, *i.e.*, if the square root of
    *op* is an integer. Under this definition both 0 and 1 are considered to
    be perfect squares.


Logical and Bit Manipulation Functions
--------------------------------------

.. table:: Logical Operators

    ======== ======================== =================================== ===================
    Operator Description              Example                             Return
    ======== ======================== =================================== ===================
    `!&`     Bitwise and              `!'0b10001'::mpz & '0b01001'::mpz`  `!'0b1'::mpz`
    `!|`     Bitwise inclusive-or     `!'0b10001'::mpz | '0b01001'::mpz`  `!'0b11001'::mpz`
    `!#`     Bitwise exclusive-or     `!'0b10001'::mpz # '0b01001'::mpz`  `!'0b11000'::mpz`
    `!-~`    One's complement         `!-~'0b10001'::mpz`                 `!'-0b10010'::mpz`
    ======== ======================== =================================== ===================
