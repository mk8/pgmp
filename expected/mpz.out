--
--  Test mpz datatype
--
--
-- first, define the datatype.  Turn off echoing so that expected file
-- does not depend on contents of pgmp.sql.
--
SET client_min_messages = warning;
\set ECHO none
RESET client_min_messages;
--
-- mpz input and output functions
--
SELECT '0'::mpz;
0
SELECT '1'::mpz;
1
SELECT '-1'::mpz;
-1
SELECT '10'::mpz;
10
SELECT '-10'::mpz;
-10
SELECT ' 1'::mpz;
1
SELECT ' 1.1'::mpz;
ERROR:  invalid input for mpz: " 1.1"
LINE 1: SELECT ' 1.1'::mpz;
               ^
SELECT ' .1'::mpz;
ERROR:  invalid input for mpz: " .1"
LINE 1: SELECT ' .1'::mpz;
               ^
SELECT ' .'::mpz;
ERROR:  invalid input for mpz: " ."
LINE 1: SELECT ' .'::mpz;
               ^
SELECT ' '::mpz;
ERROR:  invalid input for mpz: " "
LINE 1: SELECT ' '::mpz;
               ^
select mpz(' 10',8);
8
select mpz(' 10.2',8);
ERROR:  invalid input for mpz base 8: " 10.2"
select mpz(' .',8);
ERROR:  invalid input for mpz base 8: " ."
select mpz(' .2',8);
ERROR:  invalid input for mpz base 8: " .2"
SELECT '000001'::mpz;       -- padding zeros
1
SELECT '-000001'::mpz;
-1
SELECT '4294967295'::mpz;   -- limbs boundaries
4294967295
SELECT '4294967296'::mpz;
4294967296
SELECT '-4294967296'::mpz;
-4294967296
SELECT '-4294967297'::mpz;
-4294967297
SELECT '18446744073709551614'::mpz;
18446744073709551614
SELECT '18446744073709551615'::mpz;
18446744073709551615
SELECT '18446744073709551616'::mpz;
18446744073709551616
SELECT '18446744073709551617'::mpz;
18446744073709551617
SELECT '-18446744073709551615'::mpz;
-18446744073709551615
SELECT '-18446744073709551616'::mpz;
-18446744073709551616
SELECT '-18446744073709551617'::mpz;
-18446744073709551617
SELECT '-18446744073709551618'::mpz;
-18446744073709551618
SELECT '12345678901234567890123456789012345678901234567890123456789012345678901234567890'::mpz;
12345678901234567890123456789012345678901234567890123456789012345678901234567890
-- other bases
SELECT '0x10'::mpz, '010'::mpz, '0b10'::mpz;
16|8|2
SELECT mpz('10'), mpz('10', 16), mpz('10', 2), mpz('10', 62);
10|16|2|62
SELECT mpz('10', 1);
ERROR:  invalid base for mpz input: 1
HINT:  base should be between 2 and 62
SELECT mpz('10', 63);
ERROR:  invalid base for mpz input: 63
HINT:  base should be between 2 and 62
SELECT text(10::mpz);
10
SELECT text(255::mpz, 16);
ff
SELECT text((36 * 36 - 1)::mpz, 36);
zz
SELECT text((62 * 62 - 1)::mpz, 62);
zz
SELECT text((36 * 36 - 1)::mpz, -36);
ZZ
SELECT text(10::mpz, -37);
ERROR:  invalid base for mpz output: -37
HINT:  base should be between -36 and 62 and cant'be -1 or 1
SELECT text(10::mpz, 63);
ERROR:  invalid base for mpz output: 63
HINT:  base should be between -36 and 62 and cant'be -1 or 1
-- limited error
SELECT ('xx' || repeat('1234567890', 10))::mpz;
ERROR:  invalid input for mpz: "xx123456789012345678901234567890123456789012345678..."
SELECT mpz('xx' || repeat('1234567890', 10), 42);
ERROR:  invalid input for mpz base 42: "xx123456789012345678901234567890123456789012345678..."
--
-- mpz cast
--
SELECT 0::smallint::mpz, (-32768)::smallint::mpz, 32767::smallint::mpz;
0|-32768|32767
SELECT 0::integer::mpz, (-2147483648)::integer::mpz, 2147483647::integer::mpz;
0|-2147483648|2147483647
SELECT 0::bigint::mpz, (-9223372036854775808)::bigint::mpz, 9223372036854775807::bigint::mpz;
0|-9223372036854775808|9223372036854775807
SELECT 0::numeric::mpz, (-12345678901234567890)::numeric::mpz, 12345678901234567890::numeric::mpz;
0|-12345678901234567890|12345678901234567890
-- decimal are truncated
SELECT 123.10::numeric::mpz, 123.90::numeric::mpz;
123|123
SELECT (-123.10::numeric)::mpz, (-123.90::numeric)::mpz;
-123|-123
SELECT 'NaN'::numeric::mpz;
ERROR:  can't convert numeric value to mpz: "NaN"
SELECT 0::mpz, 1::mpz, (-1)::mpz;       -- automatic casts
0|1|-1
SELECT 1000000::mpz, (-1000000)::mpz;
1000000|-1000000
SELECT 1000000000::mpz, (-1000000000)::mpz;
1000000000|-1000000000
SELECT 1000000000000000::mpz, (-1000000000000000)::mpz;
1000000000000000|-1000000000000000
SELECT 1000000000000000000000000000000::mpz, (-1000000000000000000000000000000)::mpz;
1000000000000000000000000000000|-1000000000000000000000000000000
SELECT -1::mpz;       -- these take the unary minus to work
-1
SELECT -1000000::mpz;
-1000000
SELECT -1000000000::mpz;
-1000000000
SELECT -1000000000000000::mpz;
-1000000000000000
SELECT -1000000000000000000000000000000::mpz;
-1000000000000000000000000000000
SELECT 2147483647::mpz::integer;
2147483647
SELECT -2147483647::mpz::integer;
-2147483647
SELECT 2147483648::mpz::integer;
ERROR:  numeric value too big to be converted in integer data type
SELECT -2147483648::mpz::integer;
ERROR:  numeric value too big to be converted in integer data type
SELECT 32767::mpz::smallint;
32767
SELECT -32767::mpz::smallint;
-32767
SELECT 32768::mpz::smallint;
ERROR:  numeric value too big to be converted in smallint data type
SELECT -32768::mpz::smallint;
ERROR:  numeric value too big to be converted in smallint data type
SELECT 9223372036854775807::mpz::bigint;
9223372036854775807
SELECT -9223372036854775807::mpz::bigint;
-9223372036854775807
SELECT 9223372036854775808::mpz::bigint;
ERROR:  numeric value too big to be converted in biginteger data type
SELECT -9223372036854775808::mpz::bigint;
ERROR:  numeric value too big to be converted in biginteger data type
SELECT 2147483648::mpz::bigint;
2147483648
SELECT -2147483648::mpz::bigint;
-2147483648
SELECT (65536::mpz)::bigint;
65536
SELECT (65536::mpz*65536::mpz)::bigint;
4294967296
SELECT (65536::mpz*65536::mpz*65536::mpz)::bigint;
281474976710656
SELECT (65536::mpz*65536::mpz*65536::mpz*65536::mpz/2::mpz-1::mpz)::bigint;
9223372036854775807
SELECT (65536::mpz*65536::mpz*65536::mpz*65536::mpz/2::mpz)::bigint;
ERROR:  numeric value too big to be converted in biginteger data type
SELECT (-65536::mpz)::bigint;
-65536
SELECT (-65536::mpz*65536::mpz)::bigint;
-4294967296
SELECT (-65536::mpz*65536::mpz*65536::mpz)::bigint;
-281474976710656
SELECT (-65536::mpz*65536::mpz*65536::mpz*65536::mpz/2::mpz+1::mpz)::bigint;
-9223372036854775807
SELECT (-65536::mpz*65536::mpz*65536::mpz*65536::mpz/2::mpz)::bigint;
ERROR:  numeric value too big to be converted in biginteger data type
SELECT (65536::mpz)::numeric;
65536
SELECT (65536::mpz*65536::mpz)::numeric;
4294967296
SELECT (65536::mpz*65536::mpz*65536::mpz)::numeric;
281474976710656
SELECT (65536::mpz*65536::mpz*65536::mpz*65536::mpz)::numeric;
18446744073709551616
SELECT (65536::mpz*65536::mpz*65536::mpz*65536::mpz-1::mpz)::numeric;
18446744073709551615
SELECT (-65536::mpz)::numeric;
-65536
SELECT (-65536::mpz*65536::mpz)::numeric;
-4294967296
SELECT (-65536::mpz*65536::mpz*65536::mpz)::numeric;
-281474976710656
SELECT (-65536::mpz*65536::mpz*65536::mpz*65536::mpz)::numeric;
-18446744073709551616
SELECT (-65536::mpz*65536::mpz*65536::mpz*65536::mpz+1::mpz)::numeric;
-18446744073709551615
-- TODO: error on 32 bit, works on 64 bit.
-- SELECT (-65536::mpz*65536::mpz*65536::mpz*65536::mpz/2::mpz)::bigint;
SELECT (65536::mpz)::numeric;
65536
SELECT (65536::mpz*65536::mpz)::numeric;
4294967296
SELECT (65536::mpz*65536::mpz*65536::mpz)::numeric;
281474976710656
SELECT (65536::mpz*65536::mpz*65536::mpz*65536::mpz)::numeric;
18446744073709551616
SELECT (65536::mpz*65536::mpz*65536::mpz*65536::mpz-1::mpz)::numeric;
18446744073709551615
SELECT (-65536::mpz)::numeric;
-65536
SELECT (-65536::mpz*65536::mpz)::numeric;
-4294967296
SELECT (-65536::mpz*65536::mpz*65536::mpz)::numeric;
-281474976710656
SELECT (-65536::mpz*65536::mpz*65536::mpz*65536::mpz)::numeric;
-18446744073709551616
SELECT (-65536::mpz*65536::mpz*65536::mpz*65536::mpz+1::mpz)::numeric;
-18446744073709551615
--
-- mpz arithmetic
--
SELECT -('0'::mpz), +('0'::mpz), -('1'::mpz), +('1'::mpz);
0|0|-1|1
SELECT -('12345678901234567890'::mpz), +('12345678901234567890'::mpz);
-12345678901234567890|12345678901234567890
SELECT abs('-1234567890'::mpz), abs('1234567890'::mpz);
1234567890|1234567890
SELECT '1'::mpz + '2'::mpz;
3
SELECT '2'::mpz + '-4'::mpz;
-2
SELECT regexp_matches((
        ('1' || repeat('0', 1000))::mpz +
        ('2' || repeat('0', 1000))::mpz)::text,
    '^3(0000000000){100}$') IS NOT NULL;
t
SELECT '3'::mpz - '2'::mpz;
1
SELECT '3'::mpz - '5'::mpz;
-2
SELECT regexp_matches((
        ('5' || repeat('0', 1000))::mpz -
        ('2' || repeat('0', 1000))::mpz)::text,
    '^3(0000000000){100}$') IS NOT NULL;
t
SELECT '3'::mpz * '2'::mpz;
6
SELECT '3'::mpz * '-5'::mpz;
-15
SELECT regexp_matches((
        ('2' || repeat('0', 1000))::mpz *
        ('3' || repeat('0', 1000))::mpz)::text,
    '^6(00000000000000000000){100}$') IS NOT NULL;
t
-- PostgreSQL should apply the conventional precedence to operators
-- with the same name of the builtin operators.
SELECT '2'::mpz + '6'::mpz * '7'::mpz;  -- cit.
44
SELECT  '7'::mpz /  '3'::mpz;
2
SELECT '-7'::mpz /  '3'::mpz;
-2
SELECT  '7'::mpz / '-3'::mpz;
-2
SELECT '-7'::mpz / '-3'::mpz;
2
SELECT  '7'::mpz %  '3'::mpz;
1
SELECT '-7'::mpz %  '3'::mpz;
-1
SELECT  '7'::mpz % '-3'::mpz;
1
SELECT '-7'::mpz % '-3'::mpz;
-1
SELECT  '7'::mpz +/  '3'::mpz;
3
SELECT '-7'::mpz +/  '3'::mpz;
-2
SELECT  '7'::mpz +/ '-3'::mpz;
-2
SELECT '-7'::mpz +/ '-3'::mpz;
3
SELECT  '7'::mpz +%  '3'::mpz;
-2
SELECT '-7'::mpz +%  '3'::mpz;
-1
SELECT  '7'::mpz +% '-3'::mpz;
1
SELECT '-7'::mpz +% '-3'::mpz;
2
SELECT  '7'::mpz -/  '3'::mpz;
2
SELECT '-7'::mpz -/  '3'::mpz;
-3
SELECT  '7'::mpz -/ '-3'::mpz;
-3
SELECT '-7'::mpz -/ '-3'::mpz;
2
SELECT  '7'::mpz -%  '3'::mpz;
1
SELECT '-7'::mpz -%  '3'::mpz;
2
SELECT  '7'::mpz -% '-3'::mpz;
-2
SELECT '-7'::mpz -% '-3'::mpz;
-1
SELECT  '7'::mpz /  '0'::mpz;
ERROR:  division by zero
SELECT  '7'::mpz %  '0'::mpz;
ERROR:  division by zero
SELECT  '7'::mpz +/  '0'::mpz;
ERROR:  division by zero
SELECT  '7'::mpz +%  '0'::mpz;
ERROR:  division by zero
SELECT  '7'::mpz -/  '0'::mpz;
ERROR:  division by zero
SELECT  '7'::mpz -%  '0'::mpz;
ERROR:  division by zero
SELECT  '21'::mpz /! '7'::mpz;
3
SELECT  '10000000000'::mpz << 10;
10240000000000
SELECT  '10000000000'::mpz << 0;
10000000000
SELECT  '10000000000'::mpz << -1;
ERROR:  op2 can't be negative
SELECT  '1027'::mpz >>   3;
128
SELECT '-1027'::mpz >>   3;
-128
SELECT  '1027'::mpz >>  -3;
ERROR:  op2 can't be negative
SELECT  '1027'::mpz %>   3;
3
SELECT '-1027'::mpz %>   3;
-3
SELECT  '1027'::mpz %>  -3;
ERROR:  op2 can't be negative
SELECT  '1027'::mpz +>>   3;
129
SELECT '-1027'::mpz +>>   3;
-128
SELECT  '1027'::mpz +>>  -3;
ERROR:  op2 can't be negative
SELECT  '1027'::mpz +%>   3;
-5
SELECT '-1027'::mpz +%>   3;
-3
SELECT  '1027'::mpz +%>  -3;
ERROR:  op2 can't be negative
SELECT  '1027'::mpz ->>   3;
128
SELECT '-1027'::mpz ->>   3;
-129
SELECT  '1027'::mpz ->>  -3;
ERROR:  op2 can't be negative
SELECT  '1027'::mpz -%>   3;
3
SELECT '-1027'::mpz -%>   3;
5
SELECT  '1027'::mpz -%>  -3;
ERROR:  op2 can't be negative
-- power operator/functions
SELECT 2::mpz ^ 10;
1024
SELECT 2::mpz ^ 0;
1
SELECT 2::mpz ^ -1;
ERROR:  op2 can't be negative
SELECT pow(2::mpz, 10);
1024
SELECT pow(2::mpz, 0);
1
SELECT pow(2::mpz, -1);
ERROR:  op2 can't be negative
--
-- mpz ordering operators
--
select 1000::mpz =   999::mpz;
f
select 1000::mpz =  1000::mpz;
t
select 1000::mpz =  1001::mpz;
f
select 1000::mpz <>  999::mpz;
t
select 1000::mpz <> 1000::mpz;
f
select 1000::mpz <> 1001::mpz;
t
select 1000::mpz !=  999::mpz;
t
select 1000::mpz != 1000::mpz;
f
select 1000::mpz != 1001::mpz;
t
select 1000::mpz <   999::mpz;
f
select 1000::mpz <  1000::mpz;
f
select 1000::mpz <  1001::mpz;
t
select 1000::mpz <=  999::mpz;
f
select 1000::mpz <= 1000::mpz;
t
select 1000::mpz <= 1001::mpz;
t
select 1000::mpz >   999::mpz;
t
select 1000::mpz >  1000::mpz;
f
select 1000::mpz >  1001::mpz;
f
select 1000::mpz >=  999::mpz;
t
select 1000::mpz >= 1000::mpz;
t
select 1000::mpz >= 1001::mpz;
f
select mpz_cmp(1000::mpz,  999::mpz);
1
select mpz_cmp(1000::mpz, 1000::mpz);
0
select mpz_cmp(1000::mpz, 1001::mpz);
-1
-- Can create a btree index
create table test_mpz_idx (z mpz);
create index test_mpz_idx_idx on test_mpz_idx (z);
--
-- mpz aggregation
--
CREATE TABLE mpzagg(z mpz);
SELECT sum(z) FROM mpzagg;      -- NULL sum

INSERT INTO mpzagg SELECT generate_series(1, 100);
INSERT INTO mpzagg VALUES (NULL);
SELECT sum(z) FROM mpzagg;
5050
SELECT prod(z) FROM mpzagg;
93326215443944152681699238856266700490715968264381621468592963895217599993229915608941463976156518286253697920827223758251185210916864000000000000000000000000
--
-- test functions
--
SELECT mpz_test_dataset(5, 20);
88888888888888888888
88888888888888888888
88888888888888888888
88888888888888888888
88888888888888888888
SELECT numeric_test_dataset(5, '123456'::numeric);
123456
123456
123456
123456
123456
--
-- mpz functions tests
--
SELECT mpz_sqrt(25::mpz);
5
SELECT mpz_sqrt(('1' || repeat('0',100))::mpz);
100000000000000000000000000000000000000000000000000
SELECT mpz_root(25::mpz);
5
SELECT mpz_root(('1' || repeat('0',100))::mpz);
100000000000000000000000000000000000000000000000000
SELECT mpz_root(27::mpz,3);
3
SELECT mpz_root(('1' || repeat('0',100))::mpz,3);
2154434690031883721759293566519350
SELECT mpz_nextprime(5::mpz);
7
SELECT mpz_nextprime(10::mpz);
11
SELECT mpz_nextprime(100::mpz);
101
SELECT mpz_nextprime(1000::mpz);
1009
